compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java

```diff
@@ -5841,35 +5842,28 @@ private DocumentationReferenceType stringToRefType(String refTypeName) {
     private Object getIntegerLiteral(Node literal, String nodeValue, SyntaxKind sign) {
         SyntaxKind literalTokenKind = ((BasicLiteralNode) literal).literalToken().kind();
         if (literalTokenKind == SyntaxKind.DECIMAL_INTEGER_LITERAL_TOKEN) {
-            return parseLong(literal, nodeValue, nodeValue, 10, sign, DiagnosticErrorCode.INTEGER_TOO_SMALL,
-                             DiagnosticErrorCode.INTEGER_TOO_LARGE);
+            return parseLong(nodeValue, nodeValue, 10);
         } else if (literalTokenKind == SyntaxKind.HEX_INTEGER_LITERAL_TOKEN) {
             String processedNodeValue = nodeValue.toLowerCase().replace("0x", "");
-            return parseLong(literal, nodeValue, processedNodeValue, 16, sign,
-                             DiagnosticErrorCode.HEXADECIMAL_TOO_SMALL, DiagnosticErrorCode.HEXADECIMAL_TOO_LARGE);
+            return parseLong(nodeValue, processedNodeValue, 16);
         }
         return null;
     }
 
-    private Object parseLong(Node literal, String originalNodeValue,
-                             String processedNodeValue, int radix, SyntaxKind sign,
-                             DiagnosticCode code1, DiagnosticCode code2) {
+    private Object parseLong(String originalNodeValue, String processedNodeValue, int radix) {
+        // Since 9223372036854775808 is to overflow in Ballerina and -9223372036854775808 is a unary-expr where the
+        // expression is 9223372036854775808, -9223372036854775808 should also overflow even though it is
+        // within bounds for Java Long
+        if (originalNodeValue.equals(INT_MIN_OVERFLOW_VAL)) {
```

Comment:
As discussed offline, we shouldn't have to special case this here IMO. We should have modeled this as a unary expression (as opposed to how we model it as a literal atm).

For something like

```ballerina
int _ = -9223372036854775808;
```

the error should be for `9223372036854775808` and not the unary expression.

Reply:
Shall we check if doing this now is possible?

Reply:
Also related to https://github.com/ballerina-platform/ballerina-lang/issues/30020.

Reply:
```ballerina
public function main() {
    int _ = -(9223372036854775808);
}
```

used to fail as expected with `ERROR [x.bal:(2:15,2:34)] Integer '9223372036854775808' too large`.

Crashes now.

```
[2021-12-07 16:04:01,243] SEVERE {b7a.log.crash} - class java.lang.String cannot be cast to class java.lang.Long (java.lang.String and java.lang.Long are in module java.base of loader 'bootstrap') 
java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Long (java.lang.String and java.lang.Long are in module java.base of loader 'bootstrap')
        at org.wso2.ballerinalang.compiler.bir.writer.BIRInstructionWriter.visit(BIRInstructionWriter.java:329)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNonTerminator$ConstantLoad.accept(BIRNonTerminator.java:176)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRInstructionWriter.lambda$visit$1(BIRInstructionWriter.java:137)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRInstructionWriter.visit(BIRInstructionWriter.java:131)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRBasicBlock.accept(BIRNode.java:443)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRInstructionWriter.lambda$writeBBs$0(BIRInstructionWriter.java:87)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRInstructionWriter.writeBBs(BIRInstructionWriter.java:87)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRBinaryWriter.writeFunction(BIRBinaryWriter.java:297)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRBinaryWriter.lambda$writeFunctions$5(BIRBinaryWriter.java:197)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRBinaryWriter.writeFunctions(BIRBinaryWriter.java:197)
        at org.wso2.ballerinalang.compiler.bir.writer.BIRBinaryWriter.serialize(BIRBinaryWriter.java:86)
        at io.ballerina.projects.ModuleContext.cacheBIR(ModuleContext.java:490)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:460)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:389)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:159)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:137)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:113)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:182)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:112)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:90)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:203)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
```